File: src/tiposkripto/BaseSuite.ts
import { ITestResourceConfiguration, ITestArtifactory } from ".";
import { Ibdd_in_any, Ibdd_out_any } from "../CoreTypes";
import { IGivens } from "./BaseGiven";

/**
 * Represents a collection of test suites keyed by their names.
 * Suites are organized as named collections because:
 * - Tests are typically grouped into logical suites (e.g., by feature, component)
 * - Suites may have different configurations or setup requirements
 * - Named suites allow for selective test execution and better reporting
 * - This supports the hierarchical structure of test organization
 */
export type ISuites<I extends Ibdd_in_any, O extends Ibdd_out_any> = Record<
  string,
  BaseSuite<I, O>
>;

export abstract class BaseSuite<I extends Ibdd_in_any, O extends Ibdd_out_any> {
  name: string;
  givens: IGivens<I>;
  store: I["istore"];
  testResourceConfiguration: ITestResourceConfiguration;
  index: number;
  failed: boolean;
  fails: number;

  artifacts: string[] = [];

  addArtifact(path: string) {
    if (typeof path !== "string") {
      throw new Error(
        `[ARTIFACT ERROR] Expected string, got ${typeof path}: ${JSON.stringify(
          path
        )}`
      );
    }
    const normalizedPath = path.replace(/\\/g, "/"); // Normalize path separators
    this.artifacts.push(normalizedPath);
  }

  constructor(name: string, index: number, givens: IGivens<I> = {}) {
    const suiteName = name || "testSuite"; // Ensure name is never undefined
    if (!suiteName) {
      throw new Error("BaseSuite requires a non-empty name");
    }

    this.name = suiteName;
    this.index = index;
    this.givens = givens;
    this.fails = 0;
  }

  public features() {
    try {
      const features = Object.keys(this.givens)
        .map((k) => this.givens[k].features)
        .flat()
        .filter((value, index, array) => {
          return array.indexOf(value) === index;
        });
      // Convert all features to strings
      const stringFeatures = features.map((feature) => {
        if (typeof feature === "string") {
          return feature;
        } else if (feature && typeof feature === "object") {
          return feature.name || JSON.stringify(feature);
        } else {
          return String(feature);
        }
      });
      return stringFeatures || [];
    } catch (e) {
      console.error("[ERROR] Failed to extract features:", JSON.stringify(e));
      return [];
    }
  }

  public toObj() {
    const givens = Object.keys(this.givens).map((k) => {
      const givenObj = this.givens[k].toObj();
      // Ensure given features are strings
      // if (givenObj.features) {
      //   givenObj.features = givenObj.features.map(feature => {

      //     return feature;
      //     // if (typeof feature === 'string') {
      //     //   return feature;
      //     // } else if (feature && typeof feature === 'object') {
      //     //   return feature.name || JSON.stringify(feature);
      //     // } else {
      //     //   return String(feature);
      //     // }
      //   });
      // }
      return givenObj;
    });

    return {
      name: this.name,
      givens,
      fails: this.fails,
      failed: this.failed,
      features: this.features(),
      artifacts: this.artifacts
        ? this.artifacts.filter((art) => typeof art === "string")
        : [],
    };
  }

  setup(
    s: I["iinput"],
    artifactory: ITestArtifactory,
    tr: ITestResourceConfiguration
    // pm: IPM
  ): Promise<I["isubject"]> {
    console.log("mark9");

    return new Promise((res) => res(s as unknown as I["isubject"]));
  }

  assertThat(t: Awaited<I["then"]> | undefined): boolean {
    return !!t;
  }

  afterAll(store: I["istore"], artifactory: ITestArtifactory) {
    return store;
  }

  async run(
    input: I["iinput"],
    testResourceConfiguration: ITestResourceConfiguration
  ): Promise<BaseSuite<I, O>> {
    this.testResourceConfiguration = testResourceConfiguration;
    const sNdx = this.index;

    const subject = await this.setup(
      input,
      // suiteArtifactory,
      testResourceConfiguration
      // proxiedPm
    );

    for (const [gKey, g] of Object.entries(this.givens)) {
      const giver = this.givens[gKey];
      try {
        this.store = await giver.give(
          subject,
          gKey,
          testResourceConfiguration,
          this.assertThat,
          sNdx
        );
        // Add the number of failures from this given to the suite's total
        this.fails += giver.fails || 0;
      } catch (e) {
        this.failed = true;
        // Add 1 to fails for the caught error
        this.fails += 1;
        // Also add any failures from the given itself
        if (giver.fails) {
          this.fails += giver.fails;
        }
        console.error(`Error in given ${gKey}:`, e);
        // Don't re-throw to continue with other givens
      }
    }

    // Mark the suite as failed if there are any failures
    if (this.fails > 0) {
      this.failed = true;
    }

    try {
      // Ensure addArtifact is properly bound to 'this'

      this.afterAll(this.store);
    } catch (e) {
      console.error(JSON.stringify(e));
      // this.fails.push(this);
      // return this;
    }

    return this;
  }
}

File: src/lib/kafe/BaseSuite.java
package kafe;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

public abstract class BaseSuite {
    public String name;
    public Map<String, BaseGiven> givens;
    public Object store;
    public ITTestResourceConfiguration testResourceConfiguration;
    public int index;
    public boolean failed;
    public int fails;
    public List<String> artifacts;
    
    public BaseSuite(String name, Map<String, BaseGiven> givens) {
        this.name = name;
        this.givens = givens != null ? givens : new HashMap<>();
        this.artifacts = new ArrayList<>();
        this.failed = false;
        this.fails = 0;
    }
    
    public void addArtifact(String path) {
        String normalizedPath = path.replace('\\', '/');
        this.artifacts.add(normalizedPath);
    }
    
    public List<String> features() {
        List<String> features = new ArrayList<>();
        Map<String, Boolean> seen = new HashMap<>();
        
        for (BaseGiven given : givens.values()) {
            if (given.features != null) {
                for (String feature : given.features) {
                    if (!seen.containsKey(feature)) {
                        features.add(feature);
                        seen.put(feature, true);
                    }
                }
            }
        }
        return features;
    }
    
    public Map<String, Object> toObj() {
        List<Map<String, Object>> givenObjs = new ArrayList<>();
        for (BaseGiven given : givens.values()) {
            givenObjs.add(given.toObj());
        }
        
        Map<String, Object> obj = new HashMap<>();
        obj.put("name", name);
        obj.put("givens", givenObjs);
        obj.put("fails", fails);
        obj.put("failed", failed);
        obj.put("features", features());
        obj.put("artifacts", artifacts);
        return obj;
    }
    
    public abstract Object setup(Object s, Function<String, Object> artifactory, 
                                ITTestResourceConfiguration tr, Object pm);
    
    public abstract boolean assertThat(Object t);
    
    public abstract Object afterAll(Object store, Function<String, Object> artifactory, Object pm);
    
    public Object run(Object input, ITTestResourceConfiguration testResourceConfiguration,
                     Function<String, Object> artifactory, Function<String, Void> tLog, Object pm) {
        this.testResourceConfiguration = testResourceConfiguration;
        
        Function<String, Object> suiteArtifactory = (fPath) -> {
            return artifactory.apply("suite-" + index + "-" + name + "/" + fPath);
        };
        
        Object subject = setup(input, suiteArtifactory, testResourceConfiguration, pm);
        
        fails = 0;
        failed = false;
        
        for (Map.Entry<String, BaseGiven> entry : givens.entrySet()) {
            String gKey = entry.getKey();
            BaseGiven g = entry.getValue();
            try {
                store = g.give(subject, gKey, testResourceConfiguration, this::assertThat,
                              suiteArtifactory, tLog, pm, index);
                if (g.failed) {
                    fails++;
                }
            } catch (Exception e) {
                failed = true;
                fails++;
                System.err.println("Error in given " + gKey + ": " + e.getMessage());
            }
        }
        
        if (fails > 0) {
            failed = true;
        }
        
        try {
            afterAll(store, artifactory, pm);
        } catch (Exception e) {
            System.err.println("Error in after_all: " + e.getMessage());
        }
        
        return this;
    }
}
